tests/Unit/Activity/ProviderTest.php

```diff
@@ -0,0 +1,377 @@
+	public function testWrongApp() {
+		$event = $this->createMock(IEvent::class);
+		$event->expects($this->once())
+			->method('getApp')
+			->willReturn('someOtherApp');
+
+		$this->expectException(\InvalidArgumentException::class);
+		$this->provider->parse('de_DE', $event);
+	}
+
+	// Testing the parse function needs a full parse.
+	public function testFullParse() {
+		$this->l10n->expects($this->once())
+			->method('t')
+			->will($this->returnCallback(function (string $identity) {
+				return $identity;
+			}));
+
+		$user = $this->createMock(IUser::class);
+		$user->expects($this->any())
+			->method('getDisplayName')
+			->willReturn('The affected User');
+		$this->userManager->expects($this->once())
+			->method('get')
+			->with('affectedUser')
+			->willReturn($user);
+
+		$form = new Form;
+		$form->setTitle('SomeChangedNiceFormTitle');
+		$this->formMapper->expects($this->once())
+			->method('findbyHash')
+			->with('54321gfedcba')
+			->willReturn($form);
+
+		$event = $this->createMock(IEvent::class);
+		$event->expects($this->once())
+			->method('getApp')
+			->willReturn('forms');
+		$event->expects($this->atLeastOnce())
+			->method('getSubject')
+			->willReturn('newsubmission');
+		$event->expects($this->atLeastOnce())
+			->method('getSubjectParameters')
+			->willReturn([
+				'userId' => 'affectedUser',
+				'formTitle' => 'SomeNiceFormTitle',
+				'formHash' => '54321gfedcba'
+			]);
+		$event->expects($this->once())
+			->method('setParsedSubject')
+			->with('The affected User answered your form SomeChangedNiceFormTitle');
+		$event->expects($this->once())
+			->method('setRichSubject')
+			->with('{user} answered your form {formTitle}', [
+				'user' => [
+					'type' => 'user',
+					'id' => 'affectedUser',
+					'name' => 'The affected User'
+				],
+				'formTitle' => [
+					'type' => 'forms-form',
+					'id' => '54321gfedcba',
+					'name' => 'SomeChangedNiceFormTitle',
+					'link' => 'http://localhost/apps/forms/54321gfedcba/results'
+				]]);
+		$event->expects($this->once())
+			->method('setIcon')
+			->with('http://localhost/core/img/actions/add.svg');
+
+		$this->eventMerger->expects($this->once())
+			->method('mergeEvents')
+			->with('user', $event, $event)
+			->willReturn($event);
+
+		$this->assertEquals($event, $this->provider->parse('de_DE', $event, $event));
+	}
+
+	// Test right subject String
+	public function testGetSubjectString() {
+		$this->l10n->expects($this->exactly(3))
+			->method('t')
+			->will($this->returnCallback(function (string $identity) {
+				return $identity;
+			}));
+
+		$this->assertEquals('{user} has shared the form {formTitle} with you', $this->provider->getSubjectString('newshare'));
+		$this->assertEquals('{user} has shared the form {formTitle} with group {group}', $this->provider->getSubjectString('newgroupshare'));
+		$this->assertEquals('{user} answered your form {formTitle}', $this->provider->getSubjectString('newsubmission'));
```

Comment:
Use a data provider instead:
https://github.com/nextcloud/spreed/blob/master/tests/php/Activity/Provider/InvitationTest.php#L126-L141

```suggestion
	public function dataGetSubjectString() {
		return [
			['newshare', '{user} has shared the form {formTitle} with you'],
			...
		];
	}

	/**
	 * @dataProvider dataGetSubjectString
	 *
	 * @param string $subject
	 * @param string $expected
	 */
	public function testGetSubjectString(string $subject, string $expected) {
		$this->l10n->expects($this->exactly(3))
			->method('t')
			->will($this->returnCallback(function (string $identity) {
				return $identity;
			}));
		$this->assertEquals($expected, $this->provider->getSubjectString($subject));
```

Reply:
if your array ever grows big and you dont know what case 0 etc was testing for, you can use an array string key as description in new line 194, see e.g. https://github.com/nextcloud/spreed/blob/master/tests/php/Notification/NotifierTest.php#L467